RIOT-OS/RIOT
# cpu/esp32: update to Espressif's precompiled ESP32 vendor toolchain with gcc 8.4.0

### Contribution description

This PR allows the use of Espressif's precompiled ESP32x vendor toolchains (GCC 8.4.0). This means it is no longer necessary to provide an ESP32 toolchain compiled explicitly for RIOT. This is the next step required to migrate to ESP-IDF 4.4 and to support other ESP32x variants.

At the moment, the following tools are used from the precompiled ESP32x vendor toolchains:
- GCC, GDB, ... for Xtensa ESP32 and for RISC-V ESP32C3 (the RIOT port is coming soon)
- OpenOCD for all ESP32x variants
- QEMU for Xtensa ESP32

Espressif's precompiled ESP32 vendor toolchain is available for
- linux-amd64
- linux-i686
- linux-arm64
- linux-armhf
- linux-armel
- macos

so that they could be used for RIOT on all these platforms, at least theoretically (not tested).

For the migration phase, I have uploaded a [RIOT Docker build image](https://hub.docker.com/repository/docker/schorcht/riotbuild_esp32_espressif_gcc_8.4.0) that already contains Espressif's precompiled ESP32 and ESP32C3 vendor toolchains. It can be used to test this PR, PR #17601 and the the upcoming PR with the ESP32C3 port that has already been completed.

Once PR #17546 is part of the next release, the new toolchains can be merged into the official RIOT Docker build image so that the next step of migration to ESP-IDF version 4.4 (PR #17601) can take place. 

### Testing procedure

Checkout this PR and execute the following tests:

1. Local installation test:

   - Install the toolchain:
      ```
      dist/tools/esptool/install.sh all
      ```
   - Set the `PATH` variable using command:
      ```
      . dist/tools/esptools/export.sh
      ```
   - Check that all tools are found and GCC 8.4.0 is used:
      ```
      which xtensa-esp32-elf-gcc
      which openocd
      which qemu-syste-xtensa
      xtensa-esp32-elf-gcc -v
      ```
   - Compile, flash and test an application, for example:
      ```
      BOARD=esp32-wroom-32 make -C tests/shell flash term
      ```
2. Docker test

   - Pull the RIOT Docker build image:
      ```
      sudo docker pull schorcht/riotbuild_esp32_espressif_gcc_8.4.0
      ```
    - Compile an application in Docker, flash and test the application, for example:
      ```
      BUILD_IN_DOCKER=1 DOCKER="sudo docker" DOCKER_IMAGE=schorcht/riotbuild_esp32_espressif_gcc_8.4.0 \
      BOARD=esp32-wroom-32 make -C tests/shell flash term
      ```
3. QEMU and GDB local installation test:

    - Compile an application and generate the flash image for QEMU, for example:
      ```
      USEMODULE=esp32_qemu BOARD=esp32-wroom-32 make -C tests/shell flash
      ```
   - Start QEMU for ESP32.
      ```
      qemu-system-xtensa \
          -s -machine esp32 \
          -drive file=tests/shell/bin/esp32-wroom-32/qemu_flash_image.bin,if=mtd,format=raw \
          -serial tcp::5555,server,nowait
      ```
   - Use `telnet` command in a second terminal to interact with application in QEMU
      ```
      telnet localhost 5555
      ```
   - Use a third terminal to start GDB and to debug the application in QEMU
      ```
      xtensa-esp32-elf-gdb tests/shell/bin/esp32-wroom-32/tests_shell.elf
      ```
4. Test the compilation of all applications for one board
      ```
      dist/tools/compile_and_test_for_board/compile_and_test_for_board.py --no-test --jobs 16 . esp32-wroom-32
      ```
### Issues/PRs references

Prerequisite for PR #17601.
